internal/mithril-protocol-config/src/http_client/aggregator_client.rs

```diff
@@ -0,0 +1,777 @@
+use anyhow::anyhow;
+use async_trait::async_trait;
+use reqwest::header::{self, HeaderValue};
+use reqwest::{self, Client, Proxy, RequestBuilder, Response, StatusCode};
+use semver::Version;
+use slog::{Logger, debug, error, warn};
+use std::{io, sync::Arc, time::Duration};
+use thiserror::Error;
+
+use mithril_common::{
+    MITHRIL_API_VERSION_HEADER, MITHRIL_SIGNER_VERSION_HEADER, StdError,
+    api_version::APIVersionProvider,
+    entities::{ClientError, ServerError},
+    logging::LoggerExtensions,
+    messages::{AggregatorFeaturesMessage, EpochSettingsMessage},
+};
+
+const JSON_CONTENT_TYPE: HeaderValue = HeaderValue::from_static("application/json");
+
+const API_VERSION_MISMATCH_WARNING_MESSAGE: &str =
+    "OpenAPI version may be incompatible, please update your Mithril node to the latest version.";
+
+/// Error structure for the Aggregator Client.
+#[derive(Error, Debug)]
+pub enum AggregatorClientError {
+    /// The aggregator host has returned a technical error.
+    #[error("remote server technical error")]
+    RemoteServerTechnical(#[source] StdError),
+
+    /// The aggregator host responded it cannot fulfill our request.
+    #[error("remote server logical error")]
+    RemoteServerLogical(#[source] StdError),
+
+    /// Could not reach aggregator.
+    #[error("remote server unreachable")]
+    RemoteServerUnreachable(#[source] StdError),
+
+    /// Unhandled status code
+    #[error("unhandled status code: {0}, response text: {1}")]
+    UnhandledStatusCode(StatusCode, String),
+
+    /// Could not parse response.
+    #[error("json parsing failed")]
+    JsonParseFailed(#[source] StdError),
+
+    /// Mostly network errors.
+    #[error("Input/Output error")]
+    IOError(#[from] io::Error),
+
+    /// HTTP client creation error
+    #[error("HTTP client creation failed")]
+    HTTPClientCreation(#[source] StdError),
+
+    /// Proxy creation error
+    #[error("proxy creation failed")]
+    ProxyCreation(#[source] StdError),
+
+    /// No signer registration round opened yet
+    #[error("a signer registration round is not opened yet, please try again later")]
+    RegistrationRoundNotYetOpened(#[source] StdError),
```

Comment:
I'm not sure that this variant is used in this context. You can probably remove it.